StorageDriver gcs not registered when updating 2.8.1 -> 2.8.2

I tried to update a registry running in Kubernetes from version `2.8.1` to `2.8.2`. The new release image does not come up and instead fails with the following stack trace:

```
panic: StorageDriver not registered: gcs                                                                                                                                                                           
                                                                                                                                                                                                                   
goroutine 1 [running]:                                                                                                                                                                                             
github.com/docker/distribution/registry/handlers.NewApp({0xe5e8a0?, 0xc00040c180?}, 0xc0003cd880)                                                                                                                  
        github.com/docker/distribution/registry/handlers/app.go:127 +0x2411                                                                                                                                        
github.com/docker/distribution/registry.NewRegistry({0xe5e8a0?, 0xc00040c180?}, 0xc0003cd880)                                                                                                                      
        github.com/docker/distribution/registry/registry.go:161 +0x145                                                                                                                                             
github.com/docker/distribution/registry.glob..func1(0xc0000341a0?, {0xc00012d380, 0x1, 0x1})                                                                                                                       
        github.com/docker/distribution/registry/registry.go:119 +0x18f                                                                                                                                             
github.com/docker/distribution/vendor/github.com/spf13/cobra.(*Command).execute(0x1369f60, {0xc00012d340?, 0x2?, 0x2?})                                                                                            
        github.com/docker/distribution/vendor/github.com/spf13/cobra/command.go:495 +0x44f                                                                                                                         
github.com/docker/distribution/vendor/github.com/spf13/cobra.(*Command).Execute(0x136a100)                                                                                                                         
        github.com/docker/distribution/vendor/github.com/spf13/cobra/command.go:560 +0xaa                                                                                                                          
main.main()                                                                                                                                                                                                        
        github.com/docker/distribution/cmd/registry/main.go:23 +0x25 
```

Hey. Thanks for participating.

Here is the configuration including some Kubernetes commands, which hopefully illustrate the problem in more detail:

```
❯ k get secrets registry-config -o jsonpath='{.data.config\.yml}' | base64 -d
---
version: 0.1
storage:
  delete:
    enabled: true
  redirect:
    disable: false
  gcs:
    bucket: ...
    credentials: {"type": "service_account", ...}
http:
    addr: :5000
    secret: ...
auth:
  token:
    realm: https://...
    service: "token-service"
    autoredirect: false
    issuer: "docker-auth"
    rootcertbundle: /etc/docker/registry/certs/server.crt
notifications:
  endpoints: [{"name": "docker-registry-ui", "url": "http://registry-ui:8000/api/events", "headers": {"Authorization": ["Bearer ..."]}, "timeout": "1s", "threshold": 5, "backoff": "10s", "ignoredmediatypes": ["application/octet-stream"]}]

❯ k get deployments.apps registry -o yaml | grep image:
      - image: registry:2.8.1

❯ k edit deployments.apps registry 
# changing only the image version to registry:2.8.2

❯ k get po
NAME                             READY   STATUS             RESTARTS     AGE
registry-748cf4d8c7-ltm86        0/1     CrashLoopBackOff   1 (4s ago)   9s
registry-748cf4d8c7-xh7nz        0/1     CrashLoopBackOff   1 (2s ago)   6s
registry-f9bd665f4-bfqnl         1/1     Running            0            36m
registry-f9bd665f4-d9p7f         1/1     Running            0            29m

❯ k logs registry-748cf4d8c7-ltm86
panic: StorageDriver not registered: gcs

goroutine 1 [running]:
github.com/docker/distribution/registry/handlers.NewApp({0xe5e8a0?, 0xc00049f770?}, 0xc000492e00)
        github.com/docker/distribution/registry/handlers/app.go:127 +0x2411
github.com/docker/distribution/registry.NewRegistry({0xe5e8a0?, 0xc00049f770?}, 0xc000492e00)
        github.com/docker/distribution/registry/registry.go:161 +0x145
github.com/docker/distribution/registry.glob..func1(0xc0000b0170?, {0xc000406cc0, 0x1, 0x1})
        github.com/docker/distribution/registry/registry.go:119 +0x18f
github.com/docker/distribution/vendor/github.com/spf13/cobra.(*Command).execute(0x1369f60, {0xc000406c80?, 0x2?, 0x2?})
        github.com/docker/distribution/vendor/github.com/spf13/cobra/command.go:495 +0x44f
github.com/docker/distribution/vendor/github.com/spf13/cobra.(*Command).Execute(0x136a100)
        github.com/docker/distribution/vendor/github.com/spf13/cobra/command.go:560 +0xaa
main.main()
        github.com/docker/distribution/cmd/registry/main.go:23 +0x25
```

Looks like changes in ba4a6bbe02d226b08c6561cd4adf2060b5bbd09c, #3839, caused the GCS and OSS drivers to no longer implement [`storagedriver.FileWriter`](https://github.com/distribution/distribution/blob/main/registry/storage/driver/storagedriver.go#L106) - trying to build with the `BUILDTAGS` fails:

```diff
diff --git a/Dockerfile b/Dockerfile
index c6ebb789..476d138b 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -27,7 +27,7 @@ RUN --mount=type=bind,target=/src,rw \
     --mount=type=cache,target=/root/.cache/go-build \
     --mount=target=/go/pkg/mod,type=cache \
     --mount=type=bind,source=/tmp/.ldflags,target=/tmp/.ldflags,from=version \
-      set -x ; xx-go build -trimpath -ldflags "$(cat /tmp/.ldflags) ${LDFLAGS}" -o /usr/bin/registry ./cmd/registry \
+      set -x ; xx-go build -tags "${BUILDTAGS}" -trimpath -ldflags "$(cat /tmp/.ldflags) ${LDFLAGS}" -o /usr/bin/registry ./cmd/registry \
       && xx-verify --static /usr/bin/registry
 
 FROM scratch AS binary
```

Adding `_ context.Context` solves this:

```diff
diff --git a/registry/storage/driver/gcs/gcs.go b/registry/storage/driver/gcs/gcs.go
index 161fdf84..d0ca799e 100644
--- a/registry/storage/driver/gcs/gcs.go
+++ b/registry/storage/driver/gcs/gcs.go
@@ -372,7 +372,7 @@ type writer struct {
 }
 
 // Cancel removes any written content from this FileWriter.
-func (w *writer) Cancel() error {
+func (w *writer) Cancel(_ context.Context) error {
 	w.closed = true
 	err := storageDeleteObject(cloud.NewContext(dummyProjectID, w.client), w.bucket, w.name)
 	if err != nil {
```

However, this introduces a new build error:

```text
#20 [build 1/1] RUN --mount=type=bind,target=/src,rw     --mount=type=cache,target=/root/.cache/go-build     --mount=target=/go/pkg/mod,type=cache     --mount=type=bind,source=/tmp/.ldflags,target=/tmp/.ldflags,from=version       set -x ; xx-go build -tags "include_oss include_gcs" -trimpath -ldflags "$(cat /tmp/.ldflags) -s -w" -o /usr/bin/registry ./cmd/registry       && xx-verify --static /usr/bin/registry
#20 sha256:843204f621414e82fa5f704603495206a989233d4dedc26efb8a64c3ff5490e2
#20 0.224 + cat /tmp/.ldflags
#20 0.226 + xx-go build -tags 'include_oss include_gcs' -trimpath -ldflags '-X github.com/distribution/distribution/v3/version.Version=2.7.0-2189-gf7717b78.m -X github.com/distribution/distribution/v3/version.Revision=f7717b7855ca44bef5b8ccb293927a83c1d68600.m -X github.com/distribution/distribution/v3/version.Package=github.com/distribution/distribution/v3 -s -w' -o /usr/bin/registry ./cmd/registry
#20 0.550 # google.golang.org/cloud/storage
#20 0.550 vendor/google.golang.org/cloud/storage/acl.go:65:15: invalid operation: v (variable of type *"google.golang.org/api/storage/v1".ObjectAccessControl) is not an interface
#20 0.550 vendor/google.golang.org/cloud/storage/acl.go:144:15: invalid operation: v (variable of type *"google.golang.org/api/storage/v1".ObjectAccessControl) is not an interface
#20 ERROR: executor failed running [/bin/sh -c set -x ; xx-go build -tags "${BUILDTAGS}" -trimpath -ldflags "$(cat /tmp/.ldflags) ${LDFLAGS}" -o /usr/bin/registry ./cmd/registry       && xx-verify --static /usr/bin/registry]: exit code: 2
------
 > [build 1/1] RUN --mount=type=bind,target=/src,rw     --mount=type=cache,target=/root/.cache/go-build     --mount=target=/go/pkg/mod,type=cache     --mount=type=bind,source=/tmp/.ldflags,target=/tmp/.ldflags,from=version       set -x ; xx-go build -tags "include_oss include_gcs" -trimpath -ldflags "$(cat /tmp/.ldflags) -s -w" -o /usr/bin/registry ./cmd/registry       && xx-verify --static /usr/bin/registry:
------
executor failed running [/bin/sh -c set -x ; xx-go build -tags "${BUILDTAGS}" -trimpath -ldflags "$(cat /tmp/.ldflags) ${LDFLAGS}" -o /usr/bin/registry ./cmd/registry       && xx-verify --static /usr/bin/registry]: exit code: 2
```

--- a/Dockerfile
+++ b/Dockerfile
@@ -27,7 +27,7 @@ RUN --mount=type=bind,target=/go/src/github.com/docker/distribution,rw \
     --mount=type=cache,target=/root/.cache/go-build \
     --mount=target=/go/pkg/mod,type=cache \
     --mount=type=bind,source=/tmp/.ldflags,target=/tmp/.ldflags,from=version \
-      set -x ; xx-go build -trimpath -ldflags "$(cat /tmp/.ldflags) ${LDFLAGS}" -o /usr/bin/registry ./cmd/registry \
+      set -x ; xx-go build -trimpath -ldflags "$(cat /tmp/.ldflags) ${LDFLAGS}" -tags="${BUILDTAGS}" -o /usr/bin/registry ./cmd/registry \
       && xx-verify --static /usr/bin/registry
 
 FROM scratch AS binary

